lib/block_view_lockups.go

```diff
@@ -9,42 +9,129 @@ import (
 	"github.com/holiman/uint256"
 	"github.com/pkg/errors"
 	"math/big"
+	"reflect"
 	"sort"
 )
 
 //
 // TYPES: LockedBalanceEntry
 //
 
+// LockedBalanceEntry is a simple struct with different disjoint states:
+//
+//		(1) UnlockTimestampNanoSecs == VestingEndTimestampNanoSecs
+//			This is the unvested case. It represents a lump sum unlock at a given time.
+//			The user can unlock the full BalanceBaseUnits AFTER UnlockTimestampNanoSecs.
+//			When written to disk, we add a special 'byte' to the database entry key such
+//			that we can quickly find all unvested LockedBalanceEntries.
+//		(2) UnlockTimestampNanoSecs < VestingEndTimestampNanoSecs
+//			This is the vested case. It represents a vest schedule for locked DeSo tokens.
+//			The user can unlock a portion of BalanceBaseUnits proportional to
+//			how much of the time between (UnlockTimestampNanoSecs, VestingEndTimestampNanoSecs)
+//			has passed by the time the balance is unlocked. When writing this LockedBalanceEntry
+//	        to disk, we add a special 'byte' to the database entry key such that
+//	        we can quickly find all vested LockedBalanceEntries.
+//		(3) UnlockTimestampNanoSecs > VestingEndTimestampNanoSecs
+//			This shouldn't be possible. Any LockedBalanceEntries where this is the case
+//		    are degenerate and should not exist.
+//
+// These states delineate between vested and unvested LockedBalanceEntries. It's important to note that
+// by only specifying the remaining BalanceBaseUnits, UnlockTimestampNanoSecs, and VestingEndTimestampNanoSecs
+// this vesting schedule implementation is linear.
 type LockedBalanceEntry struct {
-	HODLerPKID              *PKID
-	ProfilePKID             *PKID
-	UnlockTimestampNanoSecs int64
-	BalanceBaseUnits        uint256.Int
-	isDeleted               bool
+	HODLerPKID                  *PKID
+	ProfilePKID                 *PKID
+	UnlockTimestampNanoSecs     int64
+	VestingEndTimestampNanoSecs int64
+	BalanceBaseUnits            uint256.Int
+	isDeleted                   bool
 }
 
+// LockedBalanceEntryKey is a very crucial struct in the design of lockups.
+//
+// Consider a naive utilization of LockedBalanceEntryMapKey in the context of two subsequent vested unlocks
+// WITHOUT flushing to db in-between:
+//
+//		(1st Unlock, blockTimestampNanoSecs=x+100)
+//			(1) We read from disk the LockedBalanceEntry with UnlockTimestampNanoSecs=x
+//	     	(2) We cache the entry found in-memory (i.e. in the UtxoView)
+//	     	(3) We read from the in-memory cache the LockedBalanceEntry with UnlockTimestampNanoSecs=x
+//			(4) We update the in-memory entry to have UnlockTimestampNanoSecs=x+100
```

Comment:
(3) and (4) are wrong. What we normally do is mark the entry with UnlockTimestampNanoSex=x as isDeleted=true. Otherwise, we would not know to delete it from disk on flush. Overall, I am very confused by this comment and the edge-case it's trying to describe. And I thought we were trying to avoid all this complexity by just giving each entry a unique ID.